adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,123 @@
+* We should think carefully about the right information layout, so that we get the right information in the right long-term place, and only need to do this disruptive re-arrangement once
+
+== Considered options
+
+- *Manual creation of metadata*. This is tedious for users and error-prone; we could not assume any manually-managed metadata was complete. It's also annoyingly non-DRY; we'd be asking extension owners to write the same information down in two places (their code, and a metadata file). It does have the advantage that it's non-breaking.
+- *Running builds twice*. Because the circular dependency is a soft one, it works perfectly well to run a build to generate the deployment module's files, and then run a second build to pull those into the runtime module's metadata generation. Doing this would be a simple, non-breaking, change to release scripts. However, it is error-prone (if scripts are incorrectly run once-only, we just miss metadata), lengthens build times, breaks the Develocity by introducing hidden cross-module dependencies, and is generally Just Wrong.
+- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.
+- *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.
+
+== Proposal
+
+Moving metadata generation to the 'right' place is the architecturally correct thing to do.
+
+image::images/0002-resolution.svg[the proposed resolution]
+Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg
+
+So how do we minimise the impact?
+What needs to happen for it to work?
+
+* The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module. An alternate implementation would be to publish a separate `-metadata` artifact and the tooling could read that artifact.
+* The build of extensions' deployment module invokes a maven plugin (the existing one, or a new one)
+* (optional) Extension source code refactoring:  the template `runtime/src/META-INF/quarkus-extension.yaml` moves to `deployment`. This eliminates one of the cross-module dependencies we'd like to avoid to stay cache-friendly. We want this refactoring to be optional, so if the yaml template isn't found in the deployment module, tooling would look for it in the runtime module.
```

Comment:
this here is confusing me as i feel like what is used for generation is mixed up with the actual results.

We have two things:
input to the build (i.e. src/META-INF/quarkus-extension.yaml is read as a map and then merged with values from surrounding build) which then result in a generated output.  Today we only generate one in the runtime jar.

This proposal is saying that we should have the actual generated result in the deployment.

But that doesn't require any refactoring? just adding extra step?

Reply:
> This proposal is saying that we should have the actual generated result in the deployment.
> But that doesn't require any refactoring? just adding extra step?

It depends how you define refactoring. 

Having the generated result in the deployment means tooling that looks for the generated result needs to change. The generation also has to be done 'in' the deployment jar, so extension deployment modules need to invoke `quarkus-extension-maven-plugin` in their `pom.xml`. That's a pretty invasive change, and it's the one we really care about, not the final location of the generated yaml. 

(If we weren't going to change where the generation was invoked, we'd just merge https://github.com/quarkusio/quarkus/pull/40306 as-is and accept the circular dependency.)

So that leads us to the second, optional bit. If we have to change every extension's build file so that the stuff which reads file-X is elsewhere, should we also move file-X? That feels a bit cleaner, and like stuff is logically laid out. Otherwise modules contain a file whose only purpose is to be used at build-time by another file. It also means every extension ships with two versions of file-X, one in each jar. One's enriched, one's un-enriched. That's potentially confusing. 

There's also a cache implication. Source which is logically `deployment` source is living in the `runtime` module. The consequence of having source in the 'wrong' module is that anytime file-X in the runtime jar changes, the deployment jar needs to be rebuilt. Normally, we don't have to rebuild modules just because there's a new version of one of their dependencies. So it means develocity risks getting confused and caching something it shouldn't (in the very bad case), or doing unnecessary builds (*any* change to the runtime module causes us to build the deployment module).

But there's also an argument for minimising changes, and keeping file-X where it lives now in source, because it's only used for generation, and it's location doesn't actually change anything about the actual results, apart from build times.